src/Symfony/Component/Messenger/Transport/Serialization/SerializerInterface.php

```diff
@@ -33,7 +33,7 @@ interface SerializerInterface
      *
      * @throws MessageDecodingFailedException
      */
-    public function decode(array $encodedEnvelope): Envelope;
+    public function decode(array $encodedEnvelope, ?string $contentType = null): Envelope;
```

Comment:
The content type could just be an element of the $encodedEnvelope array. Then the interface does not need to change.
Also, none of the receivers actually pass the content type to the decode method yet. So this does not seem to do anything.
If the same transport is meant to handle messages of different format (which I think is not a good design), then the doctrine transport for example would need to persist the content type as well. So that when it reads the message, it can be decoded appropriately again.

Reply:
> The content type could just be an element of the $encodedEnvelope array. Then the interface does not need to change.

Indeed, it's true, may be just adding it into the `content-type` header I added ? I may have been a bit too far on this one, I started to think about the fact that the headers and message properties are different in AMQP and that other brokers could have such distinction too and propagate conflicting information, but that's stupid of me, I'll loose that complexity.

> Also, none of the receivers actually pass the content type to the decode method yet. So this does not seem to do anything.

For now, I create the PR for discussion and feedback before going the road to update transports. AMQP will be trivial to do, Doctrine could leverage it as well by adding a `content_type` column in the storage table.

> which I think is not a good design

It's not a problem of design, as soon as you write an API or component whose goal is to speak with the outside world, you better be liberal in what you accept and conservative in what you send, and more than that, flexible in what you allow to the user of your API. Providing a dynamic content type handling is exactly the definition of being liberal, future proof, resilient to time and changes in protocol, and allowing the user of your API to manually specify the content type when (s)he sends the message is exactly being flexible and conservative: it allows a developer to target very specific nodes and comply to their capabilities.